DependencyInjection/SonataMediaExtension.php

```diff
@@ -44,6 +44,27 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('extra.xml');
         $loader->load('form.xml');
         $loader->load('gaufrette.xml');
+
+        $defS3 = $container->getDefinition('sonata.media.adapter.service.s3');
+        if (method_exists($defS3, 'setFactory')) {
+            // NEXT_MAJOR: Inline in gaufrette.xml
```

Comment:
I'm not sure I got it right, I can have in gaufrette.xml: 

```xml
    <service id="sonata.media.adapter.service.s3" class="Aws\S3\S3Client">
        <!-- NEXT_MAJOR:  Uncomment Following line
        <factory class="Aws\S3\S3Client" method="factory" />
        -->
        <argument type="collection"/>
    </service>
```

and

```xml
    <service id="sonata.media.adapter.filesystem.opencloud.objectstore" class="OpenCloud\ObjectSource\Service">
        <!-- NEXT_MAJOR: Uncomment Following line 
        <factory service="sonata.media.adapter.filesystem.opencloud.connection" method="ObjectStore" />
        -->
        <argument/>
        <argument/>
    </service>
```

and the following in SonataMediaExtension.php:

```php
    // NEXT_MAJOR: Remove Following lines
    $amazonS3Definition = $container->getDefinition('sonata.media.adapter.service.s3');
    if (method_exists($amazonS3Definition, 'setFactory')) {
        $amazonS3Definition->setFactory(array('Aws\S3\S3Client', 'factory'));
    } else {
        $amazonS3Definition->setFactoryClass('Aws\S3\S3Client');
        $amazonS3Definition->setFactoryMethod('factory');
    }

    // NEXT_MAJOR: Remove Following lines
    $openCloudDefinition = $container->getDefinition('sonata.media.adapter.filesystem.opencloud.objectstore');
    if (method_exists($openCloudDefinition, 'setFactory')) {
        $openCloudDefinition->setFactory(array(new Reference('sonata.media.adapter.filesystem.opencloud.connection'), 'ObjectStore'));
    } else {
        $openCloudDefinition->setFactoryService('sonata.media.adapter.filesystem.opencloud.connection');
        $openCloudDefinition->setFactoryMethod('ObjectStore');
    }
```